cloudformation/custom_resource/cluster.yaml

```diff
@@ -0,0 +1,225 @@
+              elif event["RequestType"].upper() == "UPDATE" and event["PhysicalResourceId"] != properties.get("ClusterName"):
+                  raise ValueError("Cannot update the ClusterName in the properties.")
+
+              physical_resource_id = properties["ClusterName"]
+
+              try:
+                  kwargs = {**{pcluster.utils.to_snake_case(k): serialize(v) for k, v in properties.items() if k not in {"ServiceToken"}}, "wait": False}
+                  func = {"CREATE": pc.create_cluster, "UPDATE": pc.update_cluster}[event["RequestType"].upper()]
+                  update_response(func(**kwargs))
+              except (APIOperationException, ParameterException, TypeError)  as e:
+                  raise ValueError(str(e))
+              except Exception as e:
+                  message = pcluster.api.errors.exception_message(e)
+                  str_msg = encoder.JSONEncoder().encode(message)
+                  if not re.search(r"No changes found", str_msg):
+                    raise ValueError(str_msg)
+
+              return physical_resource_id
+
+          @helper.create
+          def create(event, context):
+              return create_or_update(event)
+
+          @helper.update
+          def update(event, context):
+              return create_or_update(event)
+
+          @helper.delete
+          def delete(event, context):
+              cluster_name = event["ResourceProperties"].get("ClusterName")
+              try:
+                  update_response(pc.delete_cluster(cluster_name=cluster_name))
+              except (ParameterException, NotFoundException): # cluster deleted or invalid name -- ignore here.
+                  pass
+              except Exception as e:
+                  message = exception_message(e)
+                  raise ValueError(encoder.JSONEncoder().encode(message))
+
+          # Polling functionality for async CUD operations
+
+          def poll(event):
+              cluster_name = event["ResourceProperties"].get("ClusterName")
+              try:
+                  cluster = pc.describe_cluster(cluster_name=cluster_name)
+                  status = cluster.get("clusterStatus")
+
+                  if status in {"CREATE_COMPLETE", "UPDATE_COMPLETE"}:
+                      update_response(cluster)
+                      return cluster_name
+                  elif status in {"CREATE_FAILED", "UPDATE_FAILED", "DELETE_FAILED"}:
+                      raise ValueError(f"{cluster_name} failed {event['RequestType'].upper()}. See substack for further details.")
+
+              # If create fails and we try to roll-back (e.g. delete),
+              # gracefully handle missing cluster. on the delete pathway, the
+              # only invalid parameter can be the name
+              except (ParameterException, NotFoundException):
+                  if event["RequestType"].upper() == "DELETE":
+                      return True
```

Comment:
The delete poll is never supposed to return anything (the return value isn't checked) -- I removed the `True` to match this.

Reply:
Good question -- we need to return a value here to signify that polling is complete. I added a comment to signify this and change it to return `cluster_name` so that it would be the same type as the other returns.